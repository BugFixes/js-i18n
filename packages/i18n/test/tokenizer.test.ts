import tokenize from '../src/utils/tokenizer'
import type {
  Token,
} from 'moo'

type SimplifiedToken = {
  type?: string,
  value: string,
}

/**
 * Simplifies the token data to only contain useful information
 *
 * @param tokens - An array of tokens to simplify
 * @returns An array of tokens containing only the `type` and `value` properties
 */
function simplifyTokens(tokens: Token[]): SimplifiedToken[] {
  return tokens.map(function ({ type, value }) {
    return {
      type,
      value,
    }
  })
}

describe('Tokenizer', function () {
  test('correctly tokenizes a plain translation body', function () {
    const output = tokenize('The quick brown fox jumps over the lazy dog')

    expect(simplifyTokens(output)).toMatchSnapshot()
  })

  test('correctly tokenizes a translation body with an interpolated identifier', function () {
    const output = tokenize('The quick ${foxColour} fox jumps over the lazy dog') // eslint-disable-line no-template-curly-in-string

    expect(simplifyTokens(output)).toMatchSnapshot()
  })

  test('correctly tokenizes a translation body with an interpolated function call', function () {
    const output = tokenize('The quick ${getSubject()} jumps over the lazy dog') // eslint-disable-line no-template-curly-in-string

    expect(simplifyTokens(output)).toMatchSnapshot()
  })

  test('correctly tokenizes a translation body with an interpolated function call containing arguments', function () {
    const output = tokenize('The quick ${t(\'foo.bar\', { count: 2, name: \'Foobar\', locales: [\'en-GB\', \'pt-PT\'], { meta: { test: true } } })} jumps over the lazy dog') // eslint-disable-line no-template-curly-in-string

    expect(simplifyTokens(output)).toMatchSnapshot()
  })

  test('throws an error if unsupported or invalid syntax is used', function () {
    expect(function () {
      tokenize('The quick brown fox ${const foo = \'bar\'} over the lazy dog') // eslint-disable-line no-template-curly-in-string
    }).toThrow()
  })
})
