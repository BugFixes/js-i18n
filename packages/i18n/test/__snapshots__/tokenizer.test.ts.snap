// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokenizer correctly tokenizes a plain translation body 1`] = `
Array [
  Object {
    "type": "string",
    "value": "The quick brown fox jumps over the lazy dog",
  },
]
`;

exports[`Tokenizer correctly tokenizes a translation body with an interpolated function call 1`] = `
Array [
  Object {
    "type": "string",
    "value": "The quick ",
  },
  Object {
    "type": "expressionStart",
    "value": "\${",
  },
  Object {
    "type": "functionCall",
    "value": "getSubject",
  },
  Object {
    "type": "parenEnd",
    "value": ")",
  },
  Object {
    "type": "curlyEnd",
    "value": "}",
  },
  Object {
    "type": "string",
    "value": " jumps over the lazy dog",
  },
]
`;

exports[`Tokenizer correctly tokenizes a translation body with an interpolated function call containing arguments 1`] = `
Array [
  Object {
    "type": "string",
    "value": "The quick ",
  },
  Object {
    "type": "expressionStart",
    "value": "\${",
  },
  Object {
    "type": "functionCall",
    "value": "t",
  },
  Object {
    "type": "string",
    "value": "foo.bar",
  },
  Object {
    "type": "comma",
    "value": ", ",
  },
  Object {
    "type": "curlyStart",
    "value": "{ ",
  },
  Object {
    "type": "propertyName",
    "value": "count",
  },
  Object {
    "type": "number",
    "value": "2",
  },
  Object {
    "type": "comma",
    "value": ", ",
  },
  Object {
    "type": "propertyName",
    "value": "name",
  },
  Object {
    "type": "string",
    "value": "Foobar",
  },
  Object {
    "type": "comma",
    "value": ", ",
  },
  Object {
    "type": "propertyName",
    "value": "locales",
  },
  Object {
    "type": "bracketStart",
    "value": "[",
  },
  Object {
    "type": "string",
    "value": "en-GB",
  },
  Object {
    "type": "comma",
    "value": ", ",
  },
  Object {
    "type": "string",
    "value": "pt-PT",
  },
  Object {
    "type": "bracketEnd",
    "value": "]",
  },
  Object {
    "type": "comma",
    "value": ", ",
  },
  Object {
    "type": "curlyStart",
    "value": "{ ",
  },
  Object {
    "type": "propertyName",
    "value": "meta",
  },
  Object {
    "type": "curlyStart",
    "value": "{ ",
  },
  Object {
    "type": "propertyName",
    "value": "test",
  },
  Object {
    "type": "boolean",
    "value": "true",
  },
  Object {
    "type": "curlyEnd",
    "value": "} ",
  },
  Object {
    "type": "curlyEnd",
    "value": "} ",
  },
  Object {
    "type": "curlyEnd",
    "value": "}",
  },
  Object {
    "type": "parenEnd",
    "value": ")",
  },
  Object {
    "type": "curlyEnd",
    "value": "}",
  },
  Object {
    "type": "string",
    "value": " jumps over the lazy dog",
  },
]
`;

exports[`Tokenizer correctly tokenizes a translation body with an interpolated identifier 1`] = `
Array [
  Object {
    "type": "string",
    "value": "The quick ",
  },
  Object {
    "type": "expressionStart",
    "value": "\${",
  },
  Object {
    "type": "identifier",
    "value": "foxColour",
  },
  Object {
    "type": "curlyEnd",
    "value": "}",
  },
  Object {
    "type": "string",
    "value": " fox jumps over the lazy dog",
  },
]
`;
